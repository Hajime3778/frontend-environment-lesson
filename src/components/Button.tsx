import React from 'react';
import styled, { css } from 'styled-components';

const StyledButton = styled.a<ButtonStyleProps>`
  font-family: 'Nunito Sans', 'Helvetica Neue', Helvetica, Arial, sans-serif;
  font-weight: 700;
  border: 0;
  border-radius: 3em;
  cursor: pointer;
  display: inline-block;
  line-height: 1;
  text-align: center;
  width: 100px;
  ${(props: ButtonStyleProps) =>
    props.primary
      ? css`
          color: #ffffff;
          background-color: #5cc974;
        `
      : css`
          color: #333;
          background-color: transparent;
          box-shadow: rgba(0, 0, 0, 0.15) 0px 0px 0px 1px inset;
        `};
  background: ${(props: ButtonStyleProps) => props.backgroundColor};
  ${(props: ButtonStyleProps) =>
    props.size === 'small'
      ? css`
          font-size: 12px;
          padding: 10px 16px;
        `
      : props.size === 'large'
      ? css`
          font-size: 16px;
          padding: 12px 24px;
        `
      : css`
          font-size: 14px;
          padding: 11px 20px;
        `}
`;

export type ButtonStyleProps = {
  primary?: boolean;
  backgroundColor?: string;
  size?: 'small' | 'medium' | 'large';
  style?: React.CSSProperties;
};

export type ButtonProps = {
  label: string;
  onClick?: () => void;
} & ButtonStyleProps;

export const Button: React.FC<ButtonProps> = ({
  primary = false,
  size = 'medium',
  backgroundColor,
  label,
  ...props
}) => {
  return (
    <StyledButton primary={primary} size={size} backgroundColor={backgroundColor} {...props}>
      {label}
    </StyledButton>
  );
};
