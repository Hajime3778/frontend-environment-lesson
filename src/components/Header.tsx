import React from 'react';
import styled from 'styled-components';

const Title = styled.h1`
  padding: 0.25em 0 0.25em 0.75em;
  border-left: 10px solid #62d87d;
  font-size: 30px;
  font-weight: bold;
  padding: 0 0 0 10px;
`;

const RegisterButton = styled.a`
  color: #ffffff;
  background-color: #7fd989;
  width: 40px;
  height: 40px;
  text-align: center;
  line-height: 37px;
  border-radius: 50%;
  display: inline-block;
  text-decoration: none;
  transition: 0.3s;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.12), 0 2px 2px 0 rgba(0, 0, 0, 0.24);
  &:hover {
    cursor: pointer;
    color: #7fd989;
    background-color: #efefef;
    border-radius: 50%;
    text-decoration: none;
    box-shadow: 0 4px 5px 0 rgba(0, 0, 0, 0.14), 0 1px 10px 0 rgba(0, 0, 0, 0.12), 0 2px 4px -1px rgba(0, 0, 0, 0.2);
  }
`;

export type HeaderProps = {
  title: string;
  visibleRegister?: boolean;
  registerClick?: () => void;
};

export const Header: React.FC<HeaderProps> = ({ title, visibleRegister, registerClick }) => {
  return (
    <header>
      <Title>
        {title} {visibleRegister && <RegisterButton onClick={registerClick}>+</RegisterButton>}
      </Title>
    </header>
  );
};
