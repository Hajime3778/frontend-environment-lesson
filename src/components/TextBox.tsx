import React from 'react';
import styled from 'styled-components';

export type Props = {
  defaultValue?: string;
  value?: string;
  placeholder?: string;
  onChange?: (e?: React.ChangeEvent<HTMLInputElement>) => void;
};

const StyledText = styled.input`
  width: 90%;
  border-radius: 10px;
  font-size: 18px;
  border: 1px solid #dfdfdf;
  box-sizing: border-box;
  padding: 10px 15px;
  margin: 10px 0px;
  transition: 0.3s;
  &::-webkit-input-placeholder {
    color: #cacaca;
  }
  &:focus {
    outline: none;
    border: 1px solid #5cc974;
    box-shadow: 0 0 4px 0 #00802c;
  }
`;

export const TextBox: React.FC<Props> = ({ defaultValue, value, placeholder, onChange }) => {
  return (
    <StyledText type="text" defaultValue={defaultValue} value={value} placeholder={placeholder} onChange={onChange} />
  );
};
