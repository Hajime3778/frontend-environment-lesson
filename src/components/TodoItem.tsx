import React from 'react';
import { Todo } from 'src/models/Todo';

type Props = {
  todo: Todo;
  onClick: () => void;
};

export const TodoItem: React.FC<Props> = () => {
  return <div>foo</div>;
};
