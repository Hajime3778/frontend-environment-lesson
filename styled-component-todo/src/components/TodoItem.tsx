import React from 'react';
import styled from 'styled-components';
import { Todo } from 'src/models/Todo';

export type Props = {
  todo: Todo;
  onClick?: () => void;
};

const TodoArea = styled.div`
  border: 1px solid #dfdfdf;
  cursor: pointer;
  margin-bottom: 16px;
  border-radius: 8px;
  width: 100%;
  min-height: 120px;
  background-color: #fafbfd;
  box-shadow: 1px 2px 5px 3px rgba(0, 0, 0, 0.1);
  padding: 4px 2px;
`;
const TodoTitle = styled.div`
  font-size: 16px;
  font-weight: bold;
  margin: 0 10px;
  border-bottom: 1px solid #cccccc;
  text-align: left;
  padding: 8px;
`;
const TodoDescription = styled.div`
  word-wrap: break-word;
  text-align: left;
  margin-left: 10px;
  padding: 8px;
`;

export const TodoItem: React.FC<Props> = ({ todo, onClick }) => {
  return (
    <TodoArea onClick={onClick}>
      <TodoTitle>{todo.title}</TodoTitle>
      <TodoDescription>{todo.description}</TodoDescription>
    </TodoArea>
  );
};
