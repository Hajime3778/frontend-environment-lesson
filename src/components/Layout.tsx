import React from 'react';
import type { ReactNode } from 'react';
import styled from 'styled-components';
import { Header } from './Header';

export const Contents = styled.div`
  width: 800px;
  margin: 0 auto;
`;

export type LayoutProps = {
  title: string;
  visibleRegister?: boolean;
  children: ReactNode;
};

export const Layout: React.FC<LayoutProps> = (props: LayoutProps) => {
  return (
    <>
      <Contents>
        <Header title={props.title} visibleRegister={props.visibleRegister} />
        <main>{props.children}</main>
      </Contents>
    </>
  );
};
