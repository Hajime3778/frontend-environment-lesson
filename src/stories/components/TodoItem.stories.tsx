import { Story, Meta } from '@storybook/react/types-6-0';
import { Todo } from 'src/models/Todo';

import { TodoItem, Props } from 'src/components/TodoItem';

export default {
  title: 'Components/TodoItem',
  component: TodoItem,
} as Meta;

const Template: Story<Props> = (args) => <TodoItem {...args} />;

const sampleTodo: Todo = {
  id: 0,
  title: 'Sample title',
  description: 'Sample description',
};

export const Normal = Template.bind({});
Normal.args = {
  todo: sampleTodo,
};
