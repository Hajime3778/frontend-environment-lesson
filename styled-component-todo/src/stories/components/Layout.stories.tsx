import { Story, Meta } from '@storybook/react/types-6-0';
import type { ReactNode } from 'react';

import { Layout, LayoutProps } from 'src/components/Layout';

export default {
  title: 'Components/Layout',
  component: Layout,
} as Meta;

const Template: Story<LayoutProps> = (args) => <Layout {...args} />;

const sampleChild: ReactNode = <div>This is Sample</div>;

export const Normal = Template.bind({});
Normal.args = {
  title: 'Title',
  children: sampleChild,
};

export const VisibleRegister = Template.bind({});
VisibleRegister.args = {
  title: 'Title',
  visibleRegister: true,
  children: sampleChild,
};
