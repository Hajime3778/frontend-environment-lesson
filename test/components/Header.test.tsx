import { render, screen, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Header } from 'src/components/Header';

afterEach(() => cleanup());

describe('Rendering', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(<Header title="Title" />, {});
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render title only', () => {
    render(<Header title="Title" />);
    expect(screen.getByText('Title')).toBeTruthy();
    expect(screen.queryByText('+')).toBeNull();
  });

  it('Should render register button', () => {
    render(<Header title="Title" visibleRegister={true} />);
    expect(screen.getByText('Title')).toBeTruthy();
    expect(screen.getByText('+')).toBeTruthy();
  });

  it('Should trigger register button', () => {
    const mockFunction = jest.fn();
    render(<Header title="Title" visibleRegister={true} registerClick={mockFunction} />);
    userEvent.click(screen.getByText('+'));
    expect(mockFunction).toHaveBeenCalledTimes(1);
  });
});
