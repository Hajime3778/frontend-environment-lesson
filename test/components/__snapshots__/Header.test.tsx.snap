// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering Matches snapshot 1`] = `
<DocumentFragment>
  <header>
    <h1
      class="Header__Title-sc-1sdc6nh-0 enaHpz"
    >
      Title 
    </h1>
  </header>
</DocumentFragment>
`;
